### DISCUSSION TITLE
How to save part of a model for later inference: encode + decoder

### DISCUSSION THREAD

**Discussion by hmf** [deepjavalibrary/djl#567] [Q&A] [answered]
I am working on an AutoEncoder so I need to save the encoder section of the NN. 

Does anyone know of an example or have a code snippet handy that will allow me to create a Model and save the encoder part for later loading and initializing a `Predictor`? 

I am also unsure of how to implement a `Translator`

TIA

**Comment by zachgk** [comment, marked as answer]
Try making two models: one using just the encoder block that you want to save and one using the whole block that you are training with.

For the translator, it is just a class to handle whatever logic you have for your preprocessing and postprocessing. Basically, the idea is to make some convenient Java class for your input and output of your model (or use one if it already exists). Then, convert from the input to NDList in the preprocessing and NDList to output in the postprocessing.

**Reply by hmf** [reply]
> Try making two models: one using just the encoder block that you want to save and one using the whole block that you are training with.

Can those models share the block? I am assuming I can train the full model and save the encoder with the learned parameters. Is this what you mean? Now that I think of it, cant I load the model and grab only the encoder layers? 

As for the translator I think I get it.

**Reply by zachgk** [reply]
I'm pretty sure two models can share the same block. There are two options. You can either train with the full block and then save the partial block. Or, you can save the full block and then load and get the partial block.

Personally, I would go for the first option. If you just have a variable with the partial block, it should be really easy to just create a second model with it.

For the second option, but you would have to load the model, use the `Model.getBlock()`, and then get the encoder block from the full block.

**Reply by hmf** [reply]
Ok. Will try the first option. 

Thank you.
